### PR TITLE
Rework import inspection dialog in JavaFX

### PR DESCRIPTION
**Pull request by tobiasdiez**
- Rework dialog in JavaFX
- Unify code for import (move everything from the old import dialog class to the import handler)
- Simply interface and reduce user-choices (in particular, always generate bibtex keys)

There are still a few todo's related to undo/redo functionality (which is a further huge PR). Moreover, I'm not sure if I correctly transferred the handling of duplicated entries (the code was a bit hard to understand at some places).

@MartinKarim Any suggestions to (further) improve the UI? I tried to highlight the entries which the user selected for import (by clicking the plus icon), but that looked ugly and was in conflict to mouse over and mouse clicked states of the list rows.

Before:
untitled

After:
image

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/main/java/org/jabref/gui/DefaultInjector.java [hunk 1]
```diff
@@ -40,7 +42,10 @@ private static Object createDependency(Class<?> clazz) {
             return Globals.getFileUpdateMonitor();
         } else if (clazz == ClipBoardManager.class) {
             return Globals.clipboardManager;
-        } else {
+        } else if (clazz == UndoManager.class) {
+            return Globals.undoManager;
+        }
+        {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/DefaultInjector.java]: that brace position looks odd

File: src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        this.database = database;
+        this.task = task;
+
+        ViewLoader.view(this)
+                  .load()
+                  .setAsDialogPane(this);
+
+        setResultConverter(button -> {
+            if (button == importButton) {
+                viewModel.importEntries(entriesListView.getCheckModel().getCheckedItems());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java]: Is there no way to bind it to a property in the viewModel?

File: src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                        container.getChildren().add(1, duplicateButton);
+                    }
+
+                    return container;
+                })
+                .install(entriesListView);
+    }
+
+    private Node getEntryNode(BibEntry entry) {
+        Node entryType = getIcon(entry.getType()).getGraphicNode();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java]: Why don't you reuse the previewPanel? It has been designed to be used for such cases.

File: src/main/java/org/jabref/gui/importer/ImportEntriesViewModel.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    public boolean hasDuplicate(BibEntry entry) {
+        return findInternalDuplicate(entry).isPresent()
+                ||
+                DuplicateCheck.containsDuplicate(database.getDatabase(), entry, database.getMode()).isPresent();
+    }
+
+    public void importEntries(List<BibEntry> entriesToImport) {
+        // Check if we are supposed to warn about duplicates.
+        // If so, then see if there are duplicates, and warn if yes.
+        if (Globals.prefs.getBoolean(JabRefPreferences.WARN_ABOUT_DUPLICATES_IN_INSPECTION)) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/importer/ImportEntriesViewModel.java]: Inject preferencesService in the View and use it here

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 5]
```diff
@@ -667,7 +664,6 @@ private JabRefPreferences() {
         defaults.put(DO_NOT_RESOLVE_STRINGS_FOR, FieldName.URL);
         defaults.put(RESOLVE_STRINGS_ALL_FIELDS, Boolean.FALSE);
         defaults.put(NON_WRAPPABLE_FIELDS, "pdf;ps;url;doi;file;isbn;issn");
-        defaults.put(GENERATE_KEYS_AFTER_INSPECTION, Boolean.TRUE);
         defaults.put(WARN_ABOUT_DUPLICATES_IN_INSPECTION, Boolean.TRUE);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java]: Is there a reason why you removed  `GENERATE_KEYS_AFTER_INSPECTION` ?

**Comment by MartinKarim** [comment]
I would choose a simple click on the list entry the action to select the entry, which involves a change of background color. If the user then clicks on another entry, it becomes selected as well, while the first entry does not get deselected (see bottom). This would cause the whole list to be in a kind of selection mode, where clicking on an entry toggles its selection state, but does not deselect other entries in the process.
I think that mode of interaction is possible because selecting an entry is the only action that can be taken in that view if I am correct, except for clicking a button or expanding an entry.
This would have multiple advantages:

1. Changing the color on the whole entry would give the user more feedback than just coloring the circle with the plus sign.
2. It is easier to mark an entry, because the area for clicking is increased.
3. The process is easy to revert, if an entry is marked by accident.

Additionally, I would make the buttons to select/deselect entries more noticable by either making them elevated with color or by moving them to the middle of the list or even both. Right now, they are a bit overshadowed by the information above them.

importfetched

**Comment by matthiasgeiger** [comment]
I support the idea of @MartinKarim. Additionally I would suggest to have a small label above the entry list stating: "Select the entries to be imported:"

**Comment by Siedlerchr** [comment]
@MartinKarim  That is possible, Lists and Table typically support Multiple and Single selection mode.
Like in the file manager (e.g. Windows explorer), holding <kbd>shift</kbd> or pressing <kbd>CTRL</kbd> enables selection multiple entries 
I think this is also the current behavior for the main table

**Comment by matthiasgeiger** [comment]
I think the multi selection aka 'ctrl' behavior should be default without the need to hold the key pressed in order to have a good usability.

File: src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java (hunk 3 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java]: The PreviewPanel just displays HTML, which makes it hard to enrich the display with further functionality (such as the "more" button to expand the abstract). In the longterm I would actually propose to remove the PreviewPanel in favor of JavaFX controls.

File: src/main/java/org/jabref/preferences/JabRefPreferences.java (hunk 5 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java]: Yes, the keys are now always generated so there is no need to have preference toggle for it.

**Comment by tobiasdiez** [comment]
Thanks for the valuable feedback. I now implemented most of it:
image

The text buttons at the bottom still need to be improved, but since this applies to every text-based button I leave this to a new PR.

**Pull request merged**